changefeedccl: fix webhook sink retries

jayshrivastava opened the pull request:
Release note (bug fix): The `Retry.Max` field of the
`webhook_sink_config` option for changefeeds was not correctly
defined in the docs. The existing docs mention that it is "The
maximum amount of time the sink will retry a single HTTP request
to send a batch", but this is incorrect. It actually represents
the maximum number of retries which will be attempted when
sending a batch in an HTTP request fails.

Also, before this change, there was a bug where the maximum time
we would wait before retrying was 4 seconds regardless of
`Retry.Max`. With this change, the new maximum retry time is
30 seconds. We will keep doubling the initial backoff
every time we retry until the maximum of 30 seconds is reached.

Example:

If `Retry.Max = 4` and the initial backoff is 10 seconds,
at most 4 retries will be performed with the backoff times
10, 20, 30, 30 seconds respectively.

Fixes: https://github.com/cockroachdb/cockroach/issues/102467
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/102639)
<!-- Reviewable:end -->

jayshrivastava committed:
changefeedccl: fix webhook sink retries

Release note (bug fix): The `Retry.Max` field of the
`webhook_sink_config` option for changefeeds was not correctly
defined in the docs. The existing docs mention that it is "The
maximum amount of time the sink will retry a single HTTP request
to send a batch", but this is incorrect. It actually represents
the maximum number of retries which will be attempted when
sending a batch in an HTTP request fails.

Also, before this change, there was a bug where the maximum time
we would wait before retrying was 4 seconds regardless of
`Retry.Max`. With this change, the new maximum retry time is
30 seconds. We will keep doubling the initial backoff
every time we retry until the maximum of 30 seconds is reached.

Example:

If `Retry.Max = 4` and the initial backoff is 10 seconds,
at most 4 retries will be performed with the backoff times
10, 20, 30, 30 seconds respectively.

Fixes: #102467
Epic: none

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102639)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102639#-NUvgAm91jDcjREKgyHb):*
Is that so?  Because after giving up, changefeed still retires, does it not?
So, in a sense, it is max number of attempts for a single batch...?
___
*[`pkg/ccl/changefeedccl/sink_webhook.go` line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102639#-NUvgSsZ5pT6a6ONRsZu:-NUvgSsZ5pT6a6ONRsZv:b-muw5ep) ([raw file](https://github.com/cockroachdb/cockroach/blob/8597079fe734ce33471387cf7ef5f198f7627a54/pkg/ccl/changefeedccl/sink_webhook.go#L223)):*
> ```Go
> 	retryCfg.MaxRetries = int(cfg.Retry.Max)
> 	retryCfg.InitialBackoff = time.Duration(cfg.Retry.Backoff)
> 	retryCfg.MaxBackoff = retryCfg.InitialBackoff * time.Duration(int(math.Pow(2.0, float64(retryCfg.MaxRetries))))
> ```

I think we should stop doing these fancy things.  Really though, would 10 seconds or 30 seconds not be good enough for max backoff? Why would we want to it grow to hours?


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/102639)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[`-- commits` line 9 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102639#-NUvgAm91jDcjREKgyHb):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Is that so?  Because after giving up, changefeed still retires, does it not?
So, in a sense, it is max number of attempts for a single batch...?
</blockquote></details>

I updated the message to be more clear :) 
___
*[`pkg/ccl/changefeedccl/sink_webhook.go` line 223 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/102639#-NUvgSsZ5pT6a6ONRsZu:-NUvkjwG6mvjkv8XBcMX:b-187roe) ([raw file](https://github.com/cockroachdb/cockroach/blob/8597079fe734ce33471387cf7ef5f198f7627a54/pkg/ccl/changefeedccl/sink_webhook.go#L223)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think we should stop doing these fancy things.  Really though, would 10 seconds or 30 seconds not be good enough for max backoff? Why would we want to it grow to hours?
</blockquote></details>

Of course. I should have thought about that. Let's do 30 seconds. Google recommends 32 or 64. https://cloud.google.com/iot/docs/how-tos/exponential-backoff


<!-- Sent from Reviewable.io -->
